### PR TITLE
envoy: creating a stripped main base without validation server

### PR DESCRIPTION
**Pull request by alyssawilk**
Splitting main common base into a third class, to allow for a main base without the validation server

Risk Level: Medium (added fast fail for main common base validate to mittigate)
Testing: updated tests
Docs Changes: n/a
Release Notes: did not add

### CODE REVIEW

**Comment by alyssawilk** [comment]
@abeyad this one's still a draft so please do ignore for now

**Comment by jpsim** [comment]
I can confirm that this removes some stuff from the iOS framework:

```console
# Before
$ size Envoy.framework/Envoy
508M  Envoy.framework/Envoy
$ nm Envoy.framework/Envoy | grep Server14validateConfig
                 U __ZN5Envoy6Server14validateConfigERKNS0_7OptionsERKNSt3__110shared_ptrIKNS_7Network7Address8InstanceEEERNS0_16ComponentFactoryERNS_6Thread13ThreadFactoryERNS_10Filesystem8InstanceE
0000000000005548 T __ZN5Envoy6Server14validateConfigERKNS0_7OptionsERKNSt3__110shared_ptrIKNS_7Network7Address8InstanceEEERNS0_16ComponentFactoryERNS_6Thread13ThreadFactoryERNS_10Filesystem8InstanceE
00000000000c6128 b __ZZN5Envoy6Server14validateConfigERKNS0_7OptionsERKNSt3__110shared_ptrIKNS_7Network7Address8InstanceEEERNS0_16ComponentFactoryERNS_6Thread13ThreadFactoryERNS_10Filesystem8InstanceEE7flogger
000000000008c6d8 s l___func__._ZN5Envoy6Server14validateConfigERKNS0_7OptionsERKNSt3__110shared_ptrIKNS_7Network7Address8InstanceEEERNS0_16ComponentFactoryERNS_6Thread13ThreadFactoryERNS_10Filesystem8InstanceE
# After
$ size Envoy.framework/Envoy 
504M  Envoy.framework/Envoy
$ nm Envoy.framework/Envoy | grep validateConfig
```

I can do a more thorough review when it's ready.

**Comment by jmarantz** [comment]
Per IM, this requires some scrutiny of how we use the MainCommon/MainCommonBase APIs internally to ensure we have a path to being successful in the new model.

File: source/exe/main_common_base.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+class ProdComponentFactory : public Server::ComponentFactory {
+public:
+  // Server::DrainManagerFactory
+  Server::DrainManagerPtr createDrainManager(Server::Instance& server) override;
+  Runtime::LoaderPtr createRuntime(Server::Instance& server,
+                                   Server::Configuration::Initial& config) override;
+};
+
+class MainCommonBase {
```
Comment (KBaichoo) [inline source/exe/main_common_base.h:33]: Can this class be documented? e.g. especially differences between MainCommon and MainCommonBase

**Review by KBaichoo — COMMENTED** [review]
Few questions: Can you shed light on why we're splitting these e.g. what's the problem we're trying to solve? What the reason is we're removing validation from `MainCommonBase`? Thank you.

**Comment by alyssawilk** [comment]
this should now be a no-op for everything but envoy mobile.

File: source/exe/BUILD [hunk 2]
```diff
@@ -93,6 +118,14 @@ envoy_cc_library(
     ],
 )
 
+envoy_cc_library(
+    name = "envoy_stripped_main_base_lib",
+    deps = [
+        ":stripped_main_base_lib",
+        "//source/common/version:version_linkstamp",
```
Comment (abeyad) [inline source/exe/BUILD:125]: nit: from a quick glance looking at the `version_linkstamp` build target, it's not clear to me what it does, and so it's not clear to me the difference between `stripped_main_base_lib` and `envoy_stripped_main_base_lib`.. perhaps a comment would help?

File: source/exe/stripped_main_base.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+public:
+  // Server::DrainManagerFactory
+  Server::DrainManagerPtr createDrainManager(Server::Instance& server) override;
+  Runtime::LoaderPtr createRuntime(Server::Instance& server,
+                                   Server::Configuration::Initial& config) override;
+};
+
+// This is the common main between Envoy and Envoy mobile.
+// It is stripped down to functionality required by Envoy Mobile: anything
+// server-specific should live in MainCommonBase or MainCommon which remind
```
Comment (abeyad) [inline source/exe/stripped_main_base.h:35]: nit: s/remind/remain

**Review by jpsim — APPROVED** [review]
+1 I still confirmed that the validation symbols disappear from the EM binary with this change, and from what I can tell there's no functional impact to Envoy Mobile.

**Review by abeyad — APPROVED** [review]
Left some minor nits, otherwise LGTM!  It was hard to compare that the code moved from MainCommonBase to StrippedMainBase was correct, but I'm assuming the tests would catch that.

**Review by KBaichoo — APPROVED** [review]

File: source/exe/BUILD [hunk 4]
```diff
@@ -150,25 +149,21 @@ envoy_cc_library(
...
-    srcs = [
-        "main_common.cc",
-    ],
-    hdrs = [
-        "main_common.h",
-    ],
+    srcs = ["main_common.cc"],
+    hdrs = ["main_common.h"],
     deps = [
-        "//source/server/config_validation:server_lib",
```
Comment (alyssawilk) [inline source/exe/BUILD]: cc @phlax I think this version should not have passed format checks.  Did we lose buildifier on our format checks?

Comment (phlax) [reply source/exe/BUILD]: i believe it works the same as it always did

afaict the queue was cancelled before the check was run

it the `./tools/code_format/check_format.py` script that runs it i think

**Comment by alyssawilk** [comment]
@KBaichoo Ci should be sorted, PTAL!

**Review by KBaichoo — APPROVED** [review]

**Pull request merged**